docs: add new recovery.md explaning methods of fund recovery (#2933)
lightningnetwork/lnd: docs/recovery.md

# Table of Contents

> * [Recovering Funds From `lnd` (funds are safu!)](#recovering-funds-from-`lnd`-(funds-are-safu!))

Reviewer: Link should be `#recovering-funds-from-lnd-funds-are-safu`.

  * [On-Chain Recovery](#on-chain-recovery)
    * [24-word Cipher Seeds](#24-word-cipher-seeds)

>     * [Wallet+Seed Passphrases](#wallet+seed-passphrases)

Reviewer 2: TOC link is broken, along with 4 others

Reviewer: Link should be `#walletseed-passphrases`.

    * [Starting On-Chain Recovery](#starting-on-chain-recovery)
    * [Force In-Place Rescan](#force-in-place-rescan)
  * [Off-Chain Recovery](#off-chain-recovery)
    * [Obtaining SCBs](#obtaining-scbs)

>       * [On-Disk `channel.backup`](#on-disk-`channel.backup`)

Reviewer: Link should be `#on-disk-channelbackup`.

>       * [Using the `ExportChanBackup` RPC](#using-the-`exportchanbackup`-rpc)

Reviewer: Link should be `#using-the-exportchanbackup-rpc`.

>       * [Streaming Updates via `SubscribeChannelBackups`.](#streaming-updates-via-`subscribechannelbackups`.)

Reviewer: Link should be `#streaming-updates-via-subscribechannelbackups`.

    * [Recovering Using SCBs](#recovering-using-scbs)

# Recovering Funds From `lnd` (funds are safu!)

In this document, we'll go over the various built-in mechanisms for recovering
funds from `lnd` due to any sort of data loss, or malfunction. Coins in `lnd`
can exist in one of two pools: on-chain or off-chain. On-chain funds are
outputs under the control of `lnd` that can be spent immediately, and without
any auxiliary data. Off-chain funds on the other hand exist within a 2-of-2

> multi-sig output typically referred to as a payment channel. Deepening on the

Reviewer 2: s/Deepening/Depending

exact nature of operation of a given `lnd` node, one of these pools of funds
may be empty.

> Fund recovery for `lnd` will require two pieces of data: 

Reviewer 2: specify that both are only required for off-chain funds?

  1. Your 24-word cipher seed
  2. Your encrypted Static Channel Backup file (or the raw data)

The SCB file is encrypted using a key _derived_ from the user's seed. As a
result, it cannot be used in isolation.

## On-Chain Recovery 

### 24-word Cipher Seeds

When a new `lnd` node is created, it's given a 24-word seed phrase, called an

> [`aezeed`](https://github.com/lightningnetwork/lnd/tree/master/aezeed). A valid

Reviewer 2: note that it is not BIP 39, but uses the same dictionary

> see phrase obtained over the CLI `lncli create` command looks something like:

Reviewer 2: s/see/seed

```
!!!YOU MUST WRITE DOWN THIS SEED TO BE ABLE TO RESTORE THE WALLET!!!

---------------BEGIN LND CIPHER SEED---------------
 1. ability   2. noise   3. lift     4. document
 5. certain   6. month   7. shoot    8. perfect
 9. matrix   10. mango  11. excess  12. turkey
13. river    14. pitch  15. fluid   16. rack
17. drill    18. text   19. buddy   20. pool
21. soul     22. fatal  23. ship    24. jelly
---------------END LND CIPHER SEED-----------------

!!!YOU MUST WRITE DOWN THIS SEED TO BE ABLE TO RESTORE THE WALLET!!!
```

### Wallet+Seed Passphrases

During the creation process, users are first prompted to enter a **wallet
password**:
```
Input wallet password:
Confirm wallet password:
```

This password is used to _encrypt_ the wallet on disk, which includes any
derived master private keys or public key data.

> Users can also _optionally_ enter a second passphrase which we call the _cipher

Reviewer 2: above we refer to this as an "aezeed", here we call it a "cipher seed". imo we should stick to cipher seed throughout and link to the aezeed package above, what do you think?

Author: SGTM

seed passphrase_:
```
Your cipher seed can optionally be encrypted.
Input your passphrase if you wish to encrypt it (or press enter to proceed without a cipher seed passphrase):
```

If specified, then this will be used to encrypt the cipher seed itself. The
cipher seed format is unique in that the 24-word phrase is actually a
_ciphertext_. As a result, there's no standard word list as any arbitrary

> encoding can be used. If a passprhase is specified, then the cipher seed you

Reviewer 3: ```suggestion
encoding can be used. If a passphrase is specified, then the cipher seed you
```

write down is actually an _encryption_ of the entropy used to generate the BIP

> 32 root for the wallet. As a result, it can even be left in plain site if one

Reviewer 3: ```suggestion
32 root for the wallet. As a result, it can even be left in plain sight if one
```

> chooses as the passphrase is run through a standard KDF with modern parameters.

Reviewer 2: missing statement between "chooses" and "as the passphrase". tbh i don't think we should be encouraging this, even if it is relatively okay.

Unlike a BIP 39 24-word phrase, the cipher seed is able to _detect_ incorrect
passphrase. BIP 39 on the other hand, will instead silently decrypt to a new
(likely empty) wallet.

### Starting On-Chain Recovery

The initial entry point to trigger recovery of on-chain funds in the command
line is the `lncli create` command.
```

> ⛰   lncli create

Reviewer 2: remove mountain emoji?

Author: it's a terminal prompt lol

Reviewer 2: it's the only command in the doc that has it lol

```

> Next, one can enter a _new_ wallet password to encrypt any newly derived seeds

Reviewer 2: s/seeds/keys

as a result of the recovery process.
```
Input wallet password:
Confirm wallet password:
```

> nce a new wallet password has been obtained, the user will be prompted for

Reviewer 3: ```suggestion
Once a new wallet password has been obtained, the user will be prompted for
```

their _existing_ cipher seed:
```
Input your 24-word mnemonic separated by spaces: ability noise lift document certain month shoot perfect matrix mango excess turkey river pitch fluid rack drill text buddy pool soul fatal ship jelly
```

If a _cipher seed passphrase_ was used when the seed was created, it MUST be entered now:
```
Input your cipher seed passphrase (press enter if your seed doesn't have a passphrase):
```

Finally, the user has an option to choose a _recovery window_:
```
Input an optional address look-ahead used to scan for used keys (default 2500):
```

The recovery window is a metric that the on-chain rescanner will use to
determine when all the "used" addresses have been found. If the recovery window

> is 2, then after we find to consecutive addresses that haven't yet been found,

Reviewer 2: s/to/two

since the rescan doesn't actually exit on this condition, this sentence might be more accurate by saying "If the recovery window is 2, lnd will fail to find funds in any addresses generated after the point in which 2 consecutive addresses were generated but never used."

!!!YOU MUST WRITE DOWN THIS SEED TO BE ABLE TO RESTORE THE WALLET!!!

lnd successfully initialized!
```

In `lnd`'s logs, you should see something along the lines of (irrelevant lines skipped):
```
[INF] LNWL: Opened wallet
[INF] LTND: Wallet recovery mode enabled with address lookahead of 2500 addresses
[INF] LNWL: RECOVERY MODE ENABLED -- rescanning for used addresses with recovery_window=2500
[INF] CHBU: Updating backup file at test_lnd3/data/chain/bitcoin/simnet/channel.backup
[INF] CHBU: Swapping old multi backup file from test_lnd3/data/chain/bitcoin/simnet/temp-dont-use.backup to test_lnd3/data/chain/bitcoin/simnet/channel.backup
[INF] LNWL: Seed birthday surpassed, starting recovery of wallet from height=748 hash=3032830c812a4a6ea305d8ead13b52e9e69d6400ff3c997970b6f76fbc770920 with recovery-window=2500
[INF] LNWL: Scanning 1 blocks for recoverable addresses
[INF] LNWL: Recovered addresses from blocks 748-748
[INF] LNWL: Started rescan from block 3032830c812a4a6ea305d8ead13b52e9e69d6400ff3c997970b6f76fbc770920 (height 748) for 0 addresses
[INF] LNWL: Catching up block hashes to height 748, this might take a while
[INF] LNWL: Done catching up block hashes

> [INF] LNWL: Finished rescan for 0 addresses (synced to block 3032830c812a4a6ea305d8ead13b52e9e69d6400ff3c997970b6f76fbc770920, height 748)

Reviewer 2: will the number of addresses show 0?

Reviewer: Nope, it's just that the test node used for this example didn't have any addresses.

Author: filledin 800

```

That final line indicates the rescan is complete! If not all funds have
appeared, then the user may need to _repeat_ the process with a higher recovery
window. Depending on how old the wallet is (the cipher seed stores the wallet's
birthday!) and how many addresses were used, the rescan may take anywhere from
a few minutes to a few hours.

If the rescan wasn't able to complete fully (`lnd` was shutdown for example),
then from `lncli unlock`, it's possible to _restart_ the rescan from where it

> left off with the `--recovery-window` argument:

Reviewer 2: we can also note that if this argument is not provided, the wallet will not enter recovery mode and may miss funds during that portion of the rescan

```

> lncli unlock --recovery_window=250

Reviewer 2: use 2500 as before?

```

> ### Forced In-Place Rescan

Reviewer: s/Forced/Force/ to match TOC.

its birthday, or genesis if the birthday isn't known (for some older wallets).
In order to run the tool, you must first **shutdown `lnd`**. Once `lnd` is
shutdown, the rescan can be initiated with the following commands:
```
cp $HOME/.lnd/data/chain/bitcoin/mainnet/wallet.db $HOME/wallet.db # Copy the existing databse just in case!
dropwtxmgr --db=$HOME/.lnd/data/chain/bitcoin/mainnet/wallet.db
```

Once the above command returns (if it hangs for a while, then `lnd` may not
actually be shutdown, so double check!), `lnd` can be restarted. After it's
restarted, then the wallet should being rescanning. An entry resembling the
following will show up in the logs once it's complete:
```
[INF] LNWL: Finished rescan for 0 addresses (synced to block 3032830c812a4a6ea305d8ead13b52e9e69d6400ff3c997970b6f76fbc770920, height 748)
```

## Off-Chain Recovery

After version `v0.6-beta` of `lnd`, the daemon now ships with a new feature
called Static Channel Backups (SCBs). We call these _static_ as they only need
to be obtained _once_: when the channel is created. From there on, a backup is
good until the channel is closed. The backup contains all the information we

> need to initiate the Dataloss Protection (DLP) feature in the protocol, which

Reviewer: s/Dataloss/Data Loss/

ultimately leads to us recovering the funds from the channel _on-chain_. This

> is a foolproof _safe_

Reviewer 3: ```suggestion
is a foolproof _safe_ backup.
```

We say _safe_, as care has been taken to ensure that there are no foot guns in
this method of backing up channels, vs doing things like `rsync`ing or copying
the `channel.db` file periodically. Those methods can be dangerous as one never
knows if they have the latest state of a channel or not. Instead, we aim to

> provide a simple safe instead to allow users to recover the settled funds in

Reviewer 2: simple safe...

their channels in the case of partial or complete data loss. The backups
themselves are encrypted using a key derived from the user's seed, this way we
protect privacy of the users channels in the back up state, and ensure that a
random node can't attempt to import another user's channels.

Given a valid SCB, the user will be able to recover _both_ funds that are fully
settled within their channels. By "fully settled" we mean funds that are in the
base commitment outputs, and not HTLCs.  We can only restore these funds as
right after the channel is created, we have all the data required to make a

> backup. 

Reviewer 2: additionally ", but lack information about the future HTLCs that the channel will process."

### Obtaining SCBs

#### On-Disk `channel.backup`

> There're multiple ways of obtaining SCBs from `lnd`. The most commonly used

Reviewer 3: ```suggestion
There are multiple ways of obtaining SCBs from `lnd`. The most commonly used
```

method will likely be via the `channels.backup` file that's stored on-disk

> along side the rest of the chain data. This is a special file that contains SCB

Reviewer 3: ```suggestion
alongside the rest of the chain data. This is a special file that contains SCB
```

entries for _all_ currently open channels. Each time a channel is opened or

> closed, then this file is updated on disk in a safe manner (atomic file

Reviewer: s/then//

}

⛰  lncli --network=simnet exportchanbackup --all --output_file=channels.backup
```

As shown above, a user can either: specify a specific channel to backup, backup
all existing channels, or backup directly to an on-disk file. All backups use
the same format.

#### Streaming Updates via `SubscribeChannelBackups`.

Using the gRPC interace directly, [a new call:
`SubscribeChannelBackups`](https://api.lightning.community/#subscribechannelbackups).
This call allows users to receive a new notification each time the underlying

> SCB state changes. This can be used compared to the file-system approach to

Reviewer 2: Sentence needs reworking

  if a user is attempting to restore a channel due to data loss on a
  running node restored with the same seed as the node that created the
  channel. If successful, this command will allows the user to recover
  the settled funds stored in the recovered channels.

  The command will accept backups in one of three forms:

     * A single channel packed SCB, which can be obtained from
       exportchanbackup. This should be passed in hex encoded format.

     * A packed multi-channel SCB, which couples several individual
       static channel backups in single blob.

     * A file path which points to a packed multi-channel backup within a
       file, using the same format that lnd does in its channels.backup
       file.

OPTIONS:
   --single_backup value  a hex encoded single channel backup obtained from exportchanbackup
   --multi_backup value   a hex encoded multi-channel backup obtained from exportchanbackup
   --multi_file value     the path to a multi-channel back up file
```

Once the process has been initiated, `lnd` will proceed to:

  1. Given the set of channels to recover, the server will then (using the new

>      chanbackup) package, will insert a series of "channel shells" into the

Reviewer 2: link to chanbackup package

>      chanbackup) package, will insert a series of "channel shells" into the

Reviewer: Don't think we really need to mention the `chanbackup` package. Can be reworded to: 

`Given the set of channels to recover, the daemon will insert a series of "channel shells" into the database.`

>      database. These contains only the information required to initiate the DLP

Reviewer: s/contains/contain/

     (data loss protection) protocol and nothing more. As a result, they're
     marked as "recovered" channels in the database, and we'll disallow trying

>      to use then for any other process.

Reviewer: s/then/them/

>   2. Once the channel is recovered, the chanbackup package will attempt to

Reviewer 2: specify "channel shells", as opposed to just "channels?

>   2. Once the channel is recovered, the chanbackup package will attempt to

Reviewer: Same comment here w.r.t. `chanbackup` package.

     insert a LinkNode that contains all prior addresses that we were able to
     reach the peer at. During the process, we'll also insert the edge for that

>      channel (only out outgoing direction) into the database as well.

Reviewer 2: s/out/in the

  3. lnd will then start up, and as usual attempt to establish connections to

>      all peers that we have channels open with.

Reviewer 2: additionally specify what happens if lnd is already running

# Table of Contents

* [Recovering Funds From `lnd` (funds are safu!)](#recovering-funds-from-lnd-funds-are-safu)
  * [On-Chain Recovery](#on-chain-recovery)
    * [24-word Cipher Seeds](#24-word-cipher-seeds)
    * [Wallet and Seed Passphrases](#wallet-and-seed-passphrases)
    * [Starting On-Chain Recovery](#starting-on-chain-recovery)

>     * [Forced In-Place Rescan](#force-in-place-rescan)

Reviewer: Link should be `#forced-in-place-rescan`.

The initial entry point to trigger recovery of on-chain funds in the command
line is the `lncli create` command.
```
⛰   lncli create
```

Next, one can enter a _new_ wallet password to encrypt any newly derived keys
as a result of the recovery process.
```
Input wallet password:
Confirm wallet password:
```

Once a new wallet password has been obtained, the user will be prompted for
their _existing_ cipher seed:
```
Input your 24-word mnemonic separated by spaces: ability noise lift document certain month shoot perfect matrix mango excess turkey river pitch fluid rack drill text buddy pool soul fatal ship jelly
```

If a _cipher seed passphrase_ was used when the seed was created, it MUST be entered now:
```
Input your cipher seed passphrase (press enter if your seed doesn't have a passphrase):
```

Finally, the user has an option to choose a _recovery window_:
```
Input an optional address look-ahead used to scan for used keys (default 2500):
```

The recovery window is a metric that the on-chain rescanner will use to
determine when all the "used" addresses have been found. If the recovery window
is two, lnd will fail to find funds in any addresses generated after the point

> in which two consecutive addresses were generated but never used." If an `lnd`

Reviewer 2: extraneous "

 9. matrix   10. mango  11. excess  12. turkey
13. river    14. pitch  15. fluid   16. rack
17. drill    18. text   19. buddy   20. pool
21. soul     22. fatal  23. ship    24. jelly
---------------END LND CIPHER SEED-----------------

!!!YOU MUST WRITE DOWN THIS SEED TO BE ABLE TO RESTORE THE WALLET!!!

lnd successfully initialized!
```

In `lnd`'s logs, you should see something along the lines of (irrelevant lines skipped):
```
[INF] LNWL: Opened wallet
[INF] LTND: Wallet recovery mode enabled with address lookahead of 2500 addresses
[INF] LNWL: RECOVERY MODE ENABLED -- rescanning for used addresses with recovery_window=2500
[INF] CHBU: Updating backup file at test_lnd3/data/chain/bitcoin/simnet/channel.backup
[INF] CHBU: Swapping old multi backup file from test_lnd3/data/chain/bitcoin/simnet/temp-dont-use.backup to test_lnd3/data/chain/bitcoin/simnet/channel.backup
[INF] LNWL: Seed birthday surpassed, starting recovery of wallet from height=748 hash=3032830c812a4a6ea305d8ead13b52e9e69d6400ff3c997970b6f76fbc770920 with recovery-window=2500
[INF] LNWL: Scanning 1 blocks for recoverable addresses
[INF] LNWL: Recovered addresses from blocks 748-748

> [INF] LNWL: Started rescan from block 3032830c812a4a6ea305d8ead13b52e9e69d6400ff3c997970b6f76fbc770920 (height 748) for 0 addresses

Reviewer: 800 addresses*

[INF] LNWL: Catching up block hashes to height 748, this might take a while
[INF] LNWL: Done catching up block hashes
[INF] LNWL: Finished rescan for 800 addresses (synced to block 3032830c812a4a6ea305d8ead13b52e9e69d6400ff3c997970b6f76fbc770920, height 748)
```

That final line indicates the rescan is complete! If not all funds have
appeared, then the user may need to _repeat_ the process with a higher recovery
window. Depending on how old the wallet is (the cipher seed stores the wallet's
birthday!) and how many addresses were used, the rescan may take anywhere from
a few minutes to a few hours.

If the rescan wasn't able to complete fully (`lnd` was shutdown for example),
then from `lncli unlock`, it's possible to _restart_ the rescan from where it
left off with the `--recovery-window` argument:
```
⛰  lncli unlock --recovery_window=2500
```

> Note that if this argument is specified, then the wallet will not _re-enter_

Reviewer: is not*

the recovery mode and may miss funds during the portion of the rescan.

### Forced In-Place Rescan

The recovery methods described above assume a clean slate for a node, so
there's no existing UTXO or key data in the node's database. However, there're
times when an _existing_ node may want to _manually_ rescan the chain. We have
a tool for that! The tool is called
[`dropwtxmgr`](https://github.com/btcsuite/btcwallet/tree/master/cmd/dropwtxmgr).
It can be installed with the following command:
```
go get -v -u github.com/btcsuite/btcwallet/cmd/dropwtxmgr
```

The `dropwtxmgr` tool will _reset_ the best synced height of the wallet back to
its birthday, or genesis if the birthday isn't known (for some older wallets).
In order to run the tool, you must first **shutdown `lnd`**. Once `lnd` is
shutdown, the rescan can be initiated with the following commands:
```

> cp $HOME/.lnd/data/chain/bitcoin/mainnet/wallet.db $HOME/wallet.db # Copy the existing databse just in case!

Reviewer: database*

> cp $HOME/.lnd/data/chain/bitcoin/mainnet/wallet.db $HOME/wallet.db # Copy the existing databse just in case!

Reviewer 2: missing terminal prompt, and for `go get` above

After version `v0.6-beta` of `lnd`, the daemon now ships with a new feature
called Static Channel Backups (SCBs). We call these _static_ as they only need
to be obtained _once_: when the channel is created. From there on, a backup is
good until the channel is closed. The backup contains all the information we
need to initiate the Data Loss Protection (DLP) feature in the protocol, which
ultimately leads to us recovering the funds from the channel _on-chain_. This
is a foolproof _safe_ backup mechanism.

We say _safe_, as care has been taken to ensure that there are no foot guns in
this method of backing up channels, vs doing things like `rsync`ing or copying
the `channel.db` file periodically. Those methods can be dangerous as one never
knows if they have the latest state of a channel or not. Instead, we aim to
provide a simple, safe method to allow users to recover the settled funds in
their channels in the case of partial or complete data loss. The backups
themselves are encrypted using a key derived from the user's seed, this way we
protect privacy of the users channels in the back up state, and ensure that a
random node can't attempt to import another user's channels.

> Given a valid SCB, the user will be able to recover _both_ funds that are fully

Reviewer 2: _both_ out of place?
